modules/fs2/src/main/scala/com/snowplowanalytics/snowplow/enrich/fs2/Enrich.scala

```diff
@@ -90,21 +90,41 @@ object Enrich {
     _.evalTap(_ => env.metrics.badCount)
       .through(env.bad)
 
-  def goodSink[F[_]: Applicative](env: Environment[F]): ByteSink[F] =
+  def goodSink[F[_]: Applicative](env: Environment[F]): AttributedByteSink[F] =
     _.evalTap(_ => env.metrics.goodCount)
       .through(env.good)
 
-  def piiSink[F[_]: Applicative](env: Environment[F]): ByteSink[F] =
+  def piiSink[F[_]: Applicative](env: Environment[F]): AttributedByteSink[F] =
     _.through(env.pii)
 
-  def toOutputs[F[_]](result: Validated[BadRow, EnrichedEvent]): Stream[F, Output.Parsed] =
-    result match {
+  def toOutputs[F[_]](env: Environment[F]): Pipe[F, Validated[BadRow, EnrichedEvent], Output.Parsed] = {
+    val goodAttributer = addAttributes(env.goodAttributes)
+    val piiAttributer = addAttributes(env.piiAttributes)
+    _.flatMap {
       case Validated.Valid(ee) =>
-        val toEmit = List(Output.Good(ee)) ++ ConversionUtils.getPiiEvent(processor, ee).map(Output.Pii(_))
-        Stream.emits(toEmit)
+        val pii = ConversionUtils
+          .getPiiEvent(processor, ee)
+          .map(pii => Output.Pii(piiAttributer(pii)))
+        Stream.emit(Output.Good(goodAttributer(ee))) ++ Stream.emits(pii.toSeq)
       case Validated.Invalid(bad) =>
         Stream.emit(Output.Bad(bad))
     }
+  }
+
+  def addAttributes(fieldNames: Set[String]): EnrichedEvent => AttributedData[EnrichedEvent] = {
+    val fields = ConversionUtils.EnrichedFields.filter(f => fieldNames.contains(f.getName)).map(f => f.getName -> f).toMap
+
+    ee => {
+      val attrs = fields
+        .map {
+          case (k, f) => k -> Option(f.get(ee))
+        }
+        .collect {
+          case (k, Some(v)) => k -> v.toString
```

Comment:
TODO: Filter strings that exceed 1024 bytes

Reply:
I think we should trim, not filter out. Also just in case, there are very few fields that can be that long: https://github.com/snowplow/iglu-central/blob/master/schemas/com.snowplowanalytics.snowplow/atomic/jsonschema/1-0-0#L123-L157